enhancements/kube-apiserver/api-webhook-supportability.md

```diff
@@ -0,0 +1,460 @@
+  their architecture with minimal impact
+- establish an Red Hat approval process for webhooks hooking into core resources
+- establish a culture of "webhook impact budget" in the development process.
+
+### Non-Goals
+
+- tainting a cluster with effect on the support status
+- forbidding webhooks for OpenShift add-ons
+- out-of-scope: proposing alerts for kube-controller-manager, which can stopped from doing its work
+  (GC, namespace controller, quota) with unstable webhooks (especially conversion webhooks)
+- out-of-scope: creating alerts for aggregated apiservers
+
+## Proposal
+
+### User Stories
+
+1. As a cluster-admin I want to be warned if core resource webhooks risk the cluster health.
+2. As a cluster-admin I want to be warned if webhooks are slow and risk the user-experience or even
+   the availability of non-core resources (CRDs).
+3. As a cluster-admin I want to be notified and acknowledge the risk of core resource webhooks that
+   are being installed.
+4. As a development organization, we want to develop architectures that do not harm or minimally
+   harm the stability, performance, availability and security of OpenShift.
+5. As a development organization, we don't want to stop innovation and hence allow a decentralized
+   development process of OpenShift add-ons, but establish a culture where best-practices are
+   communicated (e.g. in architecture reviews) and applied.
+
+### Existing Metrics
+
+Upstream Kubernetes provides the following webhook specific metrics:
+
+- histogram `webhook_admission_duration_seconds(name, type, operation, rejected)` for per-webhook latency
+- histogram `controller_admission_duration_seconds(type, operation, rejected)` for the summed up webhook latencies
+- counter `webhook_rejection_count(name, type, operation, error_type, rejection_code)` for
+  per-webhook rejections, which includes connection errors and explicit rejection responses.
+
+Here,
+- the type is `admit` or `validating`,
+- the operation is the http verb (`CREATE`, `UPDATE`, `DELETE`, `CONNECT`),
+- the error type is `calling_webhook_error`, `apiserver_internal_error` or `no_error` (with
+  rejection code the http result code).
+
+The rejection count does not give us information about the error rate.
+
+Neither of the metrics gives us direct information about the GroupVersionResource, only about the
+admission registration object name in `webhook_admission_duration_seconds` and by that one can infer
+the affect resources. Notably, `controller_admission_duration_seconds` is not split by resource type
+in any way and hence just a cluster-global metric, and with that mostly useless.
+
+### Critical Alerts
+
+Based on the existing metrics from above, we will add the following critical alerts:
+
+- alert about latency per webhook `>1s` (`0.5s` for critical resources)
+- alert about virtual resources (SubjectAccessReview, ...) webhooks
+- alert about security sensitive webhooks (TokenReviews, oauth tokens+clients)
```

Comment:
@Anandnatraj here I would like to hear your feedback. We can decide whether this is critical or just informational.